File: go/0166_fraction_to_recurring_decimal.go
import "strconv"

func fractionToDecimal(numerator int, denominator int) string {
	flag := ""
	switch {
	case numerator < 0 && denominator < 0:
		numerator, denominator = -numerator, -denominator
	case numerator > 0 && denominator < 0:
		denominator, flag = -denominator, "-"
	case numerator < 0 && denominator > 0:
		numerator, flag = -numerator, "-"
	}

	qua, rem := numerator/denominator, numerator%denominator
	res := []byte(strconv.FormatInt(int64(qua), 10))
	if rem == 0 {
		return flag + string(res)
	}

	res = append(res, '.')
	m, curIndex := make(map[int]int), len(res)-1
	for rem != 0 {
		numerator = rem * 10
		qua, rem = numerator/denominator, numerator%denominator

		if index, ok := m[numerator]; ok {
			res = append(res, ')', ' ')
			copy(res[index+1:], res[index:])
			res[index] = '('
			break
		}

		res = append(res, digitToByte(qua))
		curIndex++
		m[numerator] = curIndex
	}
	return flag + string(res)
}

func digitToByte(digit int) byte {
	return byte('0' + digit)
}

File: python/0166_fraction_to_recurring_decimal.py
class Solution:
    def fractionToDecimal(self, numerator: int, denominator: int) -> str:
        if numerator == 0:
            return "0"

        neg = ''
        if (numerator < 0 and denominator >= 0) or (numerator >= 0 and denominator < 0):
            neg = '-'
        numerator = abs(numerator)
        denominator = abs(denominator)

        res = []

        i = -1
        while numerator >= denominator:
            d, r = numerator // denominator, numerator % denominator
            res.append(str(d))
            numerator = r
            i += 1
        if numerator == 0:
            return neg + ''.join(res)

        if not res:
            res.append('0')
            i = 0
        res.append('.')
        i += 1
        numerator *= 10

        seen = {}
        while numerator != 0:
            d, r = numerator // denominator, numerator % denominator
            if (r, d) in seen:
                res.insert(seen[(r, d)], '(')
                res.append(')')
                return neg + ''.join(res)
            else:
                res.append(str(d))
                i += 1
                seen[(r, d)] = i
                numerator = r * 10
        return neg + ''.join(res)
